app/src/main/java/io/appium/settings/location/LocationFactory.java

```diff
@@ -34,7 +36,9 @@ public synchronized Location createLocation(String providerName, float accuracy)
         l.setLatitude(latitude);
         l.setLongitude(longitude);
         l.setAltitude(altitude);
-        l.setSpeed(0);
+        if (hasSpeed) {
+            l.setSpeed(speed);
+        }
```

Comment:
should we leave the speed unset otherwise?

Reply:
Yes, setting it to a default 0.0 value was the problem. Because hasSpeed would return true even if the location did not have a proper speed value. 
I added a screenshot of the Location object from Android might be more clear.   
<img width="556" alt="Screenshot 2021-03-05 at 20 32 08" src="https://user-images.githubusercontent.com/2758798/110158328-e48d7c00-7df1-11eb-90c8-f3dc26800db0.png">

Reply:
Also I would not recommend to have a default valid value.